qualtran/bloqs/mcmt/multi_control_multi_target_pauli.py

```diff
@@ -108,50 +108,84 @@ class MultiControlPauli(GateWithRegisters):
     ) -> cirq.OP_TREE:
-        controls, target = quregs.get('controls', ()), quregs['target']
-        if len(self.cvs) < 2:
+        controls, target = quregs.get('controls', np.array([])), quregs['target']
+        if len(self.cvs) <= 2:
+            controls = controls.flatten()
+            yield [cirq.X(q) for cv, q in zip(self.cvs, controls) if cv == 0]
             yield self.target_gate.on(*target).controlled_by(*controls)
+            yield [cirq.X(q) for cv, q in zip(self.cvs, controls) if cv == 0]
             return
         qm = context.qubit_manager
         and_ancilla, and_target = np.array(qm.qalloc(len(self.cvs) - 2)), qm.qalloc(1)
-        ctrl, junk = controls[:, np.newaxis], and_ancilla[:, np.newaxis]
-        if len(self.cvs) == 2:
-            and_op = And(*self.cvs).on_registers(ctrl=ctrl, target=and_target)
-        else:
-            and_op = MultiAnd(self.cvs).on_registers(ctrl=ctrl, junk=junk, target=and_target)
+        and_op = MultiAnd(self.cvs).on_registers(
+            ctrl=controls, junk=and_ancilla[:, np.newaxis], target=and_target
+        )
         yield and_op
         yield self.target_gate.on(*target).controlled_by(*and_target)
         yield and_op**-1
         qm.qfree([*and_ancilla, *and_target])
 
     def short_name(self) -> str:
-        return r'$C^{n}(P)$'
+        n = len(self.cvs)
+        ctrl = f'C^{n}' if n > 2 else ['', 'C', 'CC'][n]
+        return f'{ctrl}{self.target_gate!s}'
 
     def _circuit_diagram_info_(self, _) -> cirq.CircuitDiagramInfo:
         wire_symbols = ["@" if b else "@(0)" for b in self.cvs]
         wire_symbols += [str(self.target_gate)]
         return cirq.CircuitDiagramInfo(wire_symbols=wire_symbols)
 
+    def on_classical_vals(self, **vals: 'ClassicalValT') -> Dict[str, 'ClassicalValT']:
+        controls, target = vals.get('controls', np.array([])), vals.get('target')
+        if self.target_gate not in (cirq.X, XGate()):
+            raise NotImplementedError(f"{self} is not classically simulatable.")
+
+        if (self.cvs == controls).all():
+            target = (target + 1) % 2
+
+        return {'controls': controls, 'target': target}
+
     def _t_complexity_(self) -> TComplexity:
-        if len(self.cvs) < 2:
-            return TComplexity(clifford=1)
-        and_gate = And(*self.cvs) if len(self.cvs) == 2 else MultiAnd(self.cvs)
-        and_cost = t_complexity(and_gate)
+        n = len(self.cvs)
+        if n <= 2:
+            pre_post_clifford = TComplexity(clifford=2 * (len(self.cvs) - sum(self.cvs)))
+            target_cost = t_complexity(self.target_gate.controlled(n))
+            return pre_post_clifford + target_cost
+        and_cost = t_complexity(MultiAnd(self.cvs))
         controlled_pauli_cost = t_complexity(self.target_gate.controlled(1))
-        and_inv_cost = t_complexity(and_gate**-1)
+        and_inv_cost = t_complexity(MultiAnd(self.cvs).adjoint())
         return and_cost + controlled_pauli_cost + and_inv_cost
```

Comment:
we should probably move this to `build_call_graph` and rely on the `from_call_graph` t complexity strategy since it's already structured this way.

Reply:
done.